keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,1072 @@
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml` . This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[ `prod-readiness-approvers` ](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+* [x] Feature gate (also fill in values in `kep.yaml`)
+  + Feature gate name: PodObservedGenerationTracking
+  + Components depending on the feature gate: Kubelet, Node Lifecycle Controller, Scheduler
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+The pod's `metadata.generation` field is currently unset by default and
+`status.observedGeneration` will be a new field in the pod status, so the feature
+will not introduce any breaking changes of default behavior.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml` .
+-->
+
+The `status.observedGeneration` feature can be disabled by setting the flag to 'false' 
+and restarting the Kubelet.
+
+The `metadata.generation` functionality will intentionally not be behind a feature gate so cannot be
```

Comment:
so when it is disabled on kubelet, it will stop setting it. Will it result in infinite loop of reconciliation as it sees that the observedGeneration is set on API server, but it is not setting it. So it wants to update it to empty, but API server doesn't allow it?

Reply:
I think you are right... let me make sure I have this right

1. the kubelet has the feature enabled, and observedGeneration is set on a pod
2. the feature is disabled on the kubelet
3. the kubelet stops setting observedGeneration, so tries to "clear" it
4. when the API server sees the observedGeneration unset in the incoming, it preserves the existing value, not letting it be cleared
5. Repeat 3-4 forever

I'm not sure what is the right solution is. Is the best thing for the API server to allow observedGeneration to be cleared? I can see this being a problem with older clients that don't know about the observedGeneration field (similar to the risk you raised about older webhooks). But seems like making sure the feature can be disabled on kubelet is more important?

Reply:
kubelet can keep setting the observedGeneration if feature gate is disabled (it will either generate it from the pod spec if FG is on or just take whatever `observedGeneration` status already have and preserve it if FG is off)

@liggitt do we have a precedence here?

Reply:
perfect, thanks for the suggestion! Moved this to beta under the graduation criteria

Reply:
> I think a reasonable path in the kubelet here is to use feature enabled || pod status has non-zero observedGeneration to decide whether to propagate metadata.generation to status.observedGeneration.

SGTM

Reply:
Depending on the complexity of the logic in kubelet to calculate the "new" value of `observedGeneration`, we may be better off by preserving the existing value if FG is disabled instead of calculating a new one. Thank you for confirming 

Reply:
I think the right way to set the new value of `observedGeneration` is to have the kubelet pass down the value of whatever `metadata.generation` is when it starts the sync loop all the way down to wherever it updates the pod status. From perusing through the source code, I don't anticipate this logic to be very complicated. So if the main concern is the complexity of the logic, I think we should be fine continuing to update `observedGeneration` after the FG is disabled rather than preserving the existing value.

Reply:
sgtm

Reply:
I folded this into the body of the KEP under "How can this feature be enabled / disabled in a live cluster?"

I'll wait for @SergeyKanzhelev's confirmation before closing this thread.

Reply:
/lgtm